rfcs/2022-08-25-12217-handling-discarded-events.md

```diff
@@ -0,0 +1,385 @@
+2. Update configuration validation to require that all outputs are handled.
+3. Add a new configuration for discarding or rejecting outputs.
+
+### User Experience
+
+#### Add Discarded Event Output
+
+We will introduce a new output to all components that would otherwise discard events, named
+`errors`. Note that some components already have such an named output. This proposal standardizes
+that output naming and provides additional support for handling it.
+
+#### Validate Handling All Outputs
+
+Additionally, we will enhance the configuration validation to determine if all outputs are
+handled. This will provide notification of the new error output and ensure users are aware of the
+new feature.
+
+In order to avoid breaking existing configurations, this validation will initially produce a
+deprecation warning, notifying the user that an output is present but unhandled. In a later version,
+this deprecation will be turned into a hard error, preventing the acceptance of a configuration with
+unhandled outputs.
+
+We will also add a new command-line option and environment variable to opt into the stricter
+validation for users that want the additional assurance this provides.
+
+#### Simplify Discarded Output Handling
+
+This enforced handling will create considerable complication for most users, as well as the
+potential of increased overhead and reduced performance if extra components have to be configured
+just to handle the outputs that now need to be connected somewhere. To reduce this overhead, we will
+add a new optional configuration setting to indicate the internal disposition of each output:
+`outputs.NAME.disposition`. This will have two non-default values, `"drop"` to mark all events going
+to that output as delivered and then drop them, and `"reject"` to mark them as having failed.  If
+this setting is not present and a component has an unhandled output, a default behavior of `"drop"`
+will be assumed, which matches the current behavior. Once validation of output handling is enforced
+as above, this default will be removed.
+
+So, to configure the new error output as discarding events with an error result, users would add the
+following:
+
+```toml
+[transforms.example]
+type = "remap"
+…
+outputs.errors.disposition = "reject"
+```
+
+### Implementation
+
+#### Configuration
+
+The new `output.*.disposition` configuration mapping will be added to components as needed to
+support configuring outputs. In order for the validation to access it, a new method will be added to
+all component configuration traits:
+
+```rust
+trait SourceConfig: NamedComponent + Debug + Send + Sync {
+    /// Gets the list of outputs exposed by this source (existing).
+    fn outputs(&self, global_log_namespace: LogNamespace) -> Vec<config::Output>;
+
+    /// The actual user configuration of outputs.
+    /// Will be `None` if no such configuration is present on this component.
+    fn outputs_config(&self) -> Option<&OutputsConfig>;
```

Comment:
`outputs_config` would lead me to believe it would also manage other aspects of outputs, for example `datadog_agent`'s `multiple_outputs` option, or it's ability to disable specific datatype streams.

Reply:
The intent is to get at the configuration described by `outputs.*.disposition`. I was leaving it as a struct to have the same pattern as the acknowledgements config.

Reply:
Yeah, no issues with that - it _may_ be worth migrating all of the "outputs config" to this struct rather than having it spread between both the component config and the new output config, if only to have a more clear mental model for the UX.

It's not directly impacting this RFC as if we need this new config struct we need it, and we don't _need_ to move the other options I mentioned. Just wanted to raise that we seemed to have a new possible "group" of config options growing here.